polygeist/lib/polygeist/Passes/AffineReduction.cpp

```diff
@@ -1,264 +1,358 @@
-#include "PassDetails.h"
+//===- AffineReduction.cpp - Affine Reduction Pass ------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
 
+#include "PassDetails.h"
 #include "mlir/Dialect/Affine/IR/AffineOps.h"
 #include "mlir/Dialect/Affine/Passes.h"
 #include "mlir/Dialect/MemRef/IR/MemRef.h"
 #include "mlir/IR/Dominance.h"
 #include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 #include "polygeist/Passes/Passes.h"
+#include "llvm/ADT/TypeSwitch.h"
 #include "llvm/Support/Debug.h"
 
+#define DEBUG_TYPE "affine-reduction"
+
 using namespace mlir;
 using namespace polygeist;
 
 namespace {
-struct AffineReductionPass : public AffineReductionBase<AffineReductionPass> {
+class AffineReductionPass : public AffineReductionBase<AffineReductionPass> {
+public:
   void runOnOperation() override;
 };
-} // end namespace.
 
-namespace {
+/// Represent a pair of affine load and store operations that satisfy all
+/// requirements for a reduction.
+class ReductionOp {
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &,
+                                       const ReductionOp &);
+
+public:
+  ReductionOp(AffineLoadOp &Load, AffineStoreOp &Store,
+              SmallVector<AffineLoadOp> &OtherLoads)
+      : Load(Load), Store(Store), OtherLoads(OtherLoads) {}
+
+  AffineLoadOp getLoad() const { return Load; }
+  AffineStoreOp getStore() const { return Store; }
+  ArrayRef<AffineLoadOp> getOtherLoads() const { return OtherLoads; }
+
+private:
+  /// The reduction load in the loop nest.
+  AffineLoadOp Load;
+  /// The reduction store in the same loop nest.
+  AffineStoreOp Store;
+  /// Compatible loads in the same loop nest.
+  SmallVector<AffineLoadOp> OtherLoads;
+};
 
-struct AffineForReductionIter : public OpRewritePattern<AffineForOp> {
+[[maybe_unused]] inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS,
+                                                      const ReductionOp &CR) {
+  OS.indent(2) << "Load: ";
+  CR.getLoad()->dump();
+  OS.indent(2) << "Store: ";
+  CR.getStore()->dump();
+  OS.indent(2) << "Other Loads: ";
+  for (const AffineLoadOp &Op : CR.getOtherLoads())
+    Op->dump();
+
+  return OS;
+}
```

Comment:
```suggestion
#ifndef NDEBUG
inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, const ReductionOp &CR) {
  OS.indent(2) << "Load: ";
  CR.getLoad()->dump();
  OS.indent(2) << "Store: ";
  CR.getStore()->dump();
  OS.indent(2) << "Other Loads: ";
  for (AffineLoadOp Op : CR.getOtherLoads())
    Op->dump();

  return OS;
}
#endif  // NDEBUG
```
I assume this is only used in debug traces. Instead of `[[maybe_unused]]`ing it (new term invented), we can wrap it around `#ifndef NDEBUG`. Also, iterate by value instead of by reference (light-weight classes).

Reply:
Yes is used in debug traces, which may be disabled. I prefer the C++17 `maybe_unused` which is part of the languaqge now (and is used throughout the LLVM code base as well). Passing a const reference is  more lightweight than passing a value, although in this case  performance doesn't matter much because this is a 'debug' aid.